stm32: add ETH support for N6 MCUs

dpgeorge opened the pull request:
### Summary

This adds support for ETH to N6 MCUs.

It's currently in draft status.  It's building and the ETH link lights work, but TX/RX do not yet work.

### Testing

Testing is done on a NUCLEO_N657X0.

### Trade-offs and Alternatives

This adds minimal code to get ETH working.  It does not affect any other stm32 MCU series, only the N6.

Unfortunately ST changed the way DMA works with the ETH peripheral, so this is a little more involved and messy than I would have liked.

dpgeorge committed:
stm32/eth: Add support for Ethernet on N6 MCUs.

This adds working support for 100MBit Ethernet on N6 MCUs.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/boards/NUCLEO_N657X0: Enable Ethernet.

Signed-off-by: Damien George <damien@micropython.org>

kwagyeman commented on ports/stm32/eth.c:
```diff
@@ -197,22 +216,22 @@ int eth_init(eth_t *self, int mac_idx, uint32_t phy_addr, int phy_type) {
+    #elif defined(STM32H7) || defined(STM32N6)
     __HAL_RCC_ETH1MAC_CLK_ENABLE();
     __HAL_RCC_ETH1TX_CLK_ENABLE();
     __HAL_RCC_ETH1RX_CLK_ENABLE();
```
Might need to enable sleep clocks for the peripheral.

kwagyeman replied:
Yeah, just asssume you have to enable sleep clocks always for all pheripherals on the N6 all the time...

dpgeorge committed:
stm32/boards/OPENMV_N6: Enable RGMII Ethernet.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/network_lan: Allow a board to configure the default LAN PHY.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/eth_phy: Add support for RTL8211 ETH PHY.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/eth: Add support for gigabit RGMII peripheral interface.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/eth_phy: Add support for 1000Mbit PHY.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
This PR is now complete and should be ready to merge:
- cleaned up commits
- full support for RMII and RGMII interfaces
- auto-detect of PHY link speed (10/100/1000MBit)
- ran all networking tests and networking multitests on OPENMV_N6 board, and they all pass

@iabdalkader would you like to test this PR?

kwagyeman commented:
Testing now.

kwagyeman commented:
On 1000 Mb/s Ethernet Link:

https://github.com/user-attachments/assets/07be593b-4823-4e9b-9392-ec27ca653451

MicroPython is Server:
TCP Receive from PC: --- Failed (might be my test script by iperf3 is failing? Can you check)
TCP Send to PC: 12.2 Mbits/sec
UDP Send to PC: 115 Mbits/sec

PC is Server:
TCP Receive from PC: --- Failed (might be my test script by iperf3 is failing? Can you check)
TCP Send to PC: 12.2 Mbits/sec
UDP Send to PC: 10.5 Mbits/sec

kwagyeman commented:
Our MPU region allocation code allocates a general DMA region, which we put all DMA buffers into. Then the general-purpose memory is located in the DTCM. Without `#define MICROPY_HW_ETH_DMA_ATTRIBUTE        __attribute__((aligned(16384), section(".dma_buffer")));` the H7 DMA buffer ends up in the DTCM, which doesn't work. This defines relocates it to the DMA buffer region, but then two MPU entries are created. The first one is by our general-purpose code, and the second one is by the ETH driver when it starts.

This isn't an error on the V7 MPU but causes an error on the V8 MPU. However, generally, we shouldn't be adding overlapping entries.

iabdalkader commented:
> the issue with two overlapping MPU regions on the H7 is not fixed

There's no issue here. Higher regions number have higher priority when regions overlapped, so ours takes precedence.

andrewleech commented:
FWIW I've recently been doing a lot of rework and testing on the stm32 eth driver in https://github.com/micropython/micropython/pull/17613 to improve the user experience regarding setting ip address at any time, fixing startup so you can .active(True) without a cable plugged in and ensuring link status is always correct. It also improves handling of dhcp addresses when hot plugging the ethernet cable.

I've only been testing this on a H5 and F4 Nucleo's but it shouldn't be too much work for me to merge the changes here into that if you'd like to test these updates on the N6 hardware too?

dpgeorge committed:
stm32/lwip_inc: Include HAL header to get MCU define.

Otherwise the custom memory settings don't take place.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/eth_phy: Move PHY initialization to a dedicated function.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/eth_phy: Add support for RTL8211 inititialization.

This gets the status indicator working on the RTL8211.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/eth: Make TX and RX buffer sizes a multiple of 8.

Needed by N6, and doesn't hurt for other MCUs.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
> 3\. The status LED is not implemented.

I have now implemented the PHY status LED.  It indicates 10/100/1000 link speed and activity.


> * I'm getting some ipref issues at the 1gig link speed. RTSP streaming worked fine though. Please check.
> 
> * iperf speeds for 1g leave room for improvement. Would it be easy to get the speed up to 800 Mb/s with UDP? I was able to hit this with the ST DK IPREF example project. 100M speed at 80Mb/s UDP is okay.

I have made some minor fixes which may improve both the reliability and throughput.  At least with my testing on OpenMV N6 the TCP iperf3 test is more reliable (in both directions), and the performance is better, up to around 23Mbit/sec.

@kwagyeman when you achieved 800 Mbit/sec UDP, was that with zero (or close to zero) packet loss?  Also which direction?  Did you run iperf3 on the host PC for that test?

kwagyeman commented:
@dpgeorge - It was 800 Mb/s send to the PC from the camera. Packet loss doesn't really matter, just the rate at which the camera records after ipref finishes. So, I'm looking to hit that kinda of rate for releasing packets. If it's an issue with MP overhead then... not much we can do there. However, it may just be a need for more packet buffers.

kwagyeman commented:
@dpgeorge - Also, feel free to massively increase the lwip buffers, dma buffers in the driver, and etc. We have plenty of RAM on the N6.

dpgeorge commented:
> It was 800 Mb/s send to the PC from the camera. Packet loss doesn't really matter, just the rate at which the camera records after ipref finishes

OK, so the main thing there is how fast the N6 can send out UDP packets, back to back.

I did some tests for this, just a simple Python script (running on the N6) which sent out UDP packets in a loop as fast as possible.  With very large UDP packets (30kbytes) I could get nearly 500MBit/sec as measured by the N6 itself (simply time how long it took to call `socket.sendto`).

I then also wrote a simple Python script to run on the host PC, to receive this stream of UDP packets.  It was not that reliable (not sure if it's my local network) but sometimes it could receive most packets and measure a rate of around 400MBit/sec.

Then, playing around with `iperf3.py`, I found that the limiting factor with that is simple Python execution speed.  The UDP rate is low with `iperf3.py` simply because it can't send out packets fast enough to keep the gigabit link saturated.  I could fix this a little by increasing the UDP packet size to 10's of kbytes, and then get around 200MBit/sec.

So, to get high performance you'll need some fast C code and/or big UDP packets (on the N6 side).

kwagyeman commented:
@dpgeorge Can you push the script to the limit to see what is possible?

kwagyeman commented:
Okay, will get to it on Friday. Please make any TCP buffers and etc in LWIP massive.

Note that being able to stream data quickly means in the future we can connect OpenMV IDE to the camera other Ethernet and it's also very important for the OpenMV interfacing in ROS, as an IP camera, and etc. there are a lot of high value use cases with Ethernet.

Like, Ethernet will be how the OpenMV Cam interfaces with large robotic systems and etc. most likely.

dpgeorge commented:
@andrewleech I tested your branch (checked out and built as-is) and deployed to the OPENMV_N6.  `lan.active(True)` returns after about 100ms, but then nothing happens.  Using `lan.config(trace=7)` I don't see any packets coming in or going out. It never gets an IP address.

andrewleech commented:
> I've built your latest code, here is the output

Thanks, looks like I missed some of the clock setup, should have checked that closer as thost sleep clocks were the thing I spent the longest manually testing on my branch! Updated now if you're still keen.

kwagyeman commented:
@dpgeorge it ran from sram. I just loaded the program via the debugger. I can send a link to the source in a bit.

dpgeorge committed:
stm32/lwip_inc: Further increase N6 lwIP memory.

This MCU has a lot of memory and gigabit ethernet, so it benefits from
improved lwIP performance.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/eth: Implement zero-copy of lwIP pbufs for TX path.

This option (currently only enabled for N6) allows the TX path to hold on
to a pbuf reference while the DMA accesses the pbuf's memory directly,
instead of copying the entire pbuf data into the internal buffers.

This is necessary to achieve gigabit speeds on the N6, although actually
achieving that speed requires higher up parts of the stack to be efficient
as well.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/eth: Don't pad eth_dma_t struct on H5 and N6 MCUs.

It's not necessary (and wastes RAM), the MPU can handle regions of
arbitrary size.

Signed-off-by: Damien George <damien@micropython.org>

andrewleech commented:
Thanks, I suspected it'd be something like that. To test raw hardware performance I was thinking the STM TCP/UDP echo server examples could be a good reference but I haven't reviewed whether they put much effort into dma / buffer optimisations or if it's bare minimum impl.

kwagyeman commented:
@dpgeorge - Great job! Looks like it peaked at 630Mbits too for me:

From the camera:

```
>>> 100 93.96286 Mbit/s
200 181.7149 Mbit/s
300 242.69392 Mbit/s
400 332.3294 Mbit/s
500 419.6832 Mbit/s
600 488.5993 Mbit/s
700 567.146 Mbit/s
800 599.4194 Mbit/s
900 609.24 Mbit/s
1000 626.125 Mbit/s
1100 632.0024 Mbit/s
1200 636.2672 Mbit/s
1300 637.1768 Mbit/s
1400 641.1358 Mbit/s
1500 393.4168 Mbit/s
1600 400.9774 Mbit/s
1700 408.6538 Mbit/s
1800 421.4963 Mbit/s
1900 447.6908 Mbit/s
2000 421.574 Mbit/s
3000 466.2911 Mbit/s
4000 517.13 Mbit/s
5000 542.3948 Mbit/s
6000 508.7494 Mbit/s
7000 542.0054 Mbit/s
8000 544.579 Mbit/s
9000 527.5342 Mbit/s
10000 541.3564 Mbit/s
11000 545.4354 Mbit/s
12000 557.88 Mbit/s
13000 561.5824 Mbit/s
14000 562.4496 Mbit/s
15000 566.4144 Mbit/s
16000 570.2194 Mbit/s
17000 565.021 Mbit/s
18000 558.7546 Mbit/s
19000 562.5254 Mbit/s
```

And on PC:

```
$ python -u udp_perf.py
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56733), packet_size 100
1000 100000 7738
100.0 %   103.3858878263117 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56734), packet_size 200
1000 200000 15847
100.0 %   100.96548242569571 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56735), packet_size 300
1000 300000 14863
100.0 %   161.4748032025836 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56736), packet_size 400
1000 400000 10771
100.0 %   297.09404883483427 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56737), packet_size 500
847 423500 8223
84.7 %   412.01507965462724 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56738), packet_size 600
1000 600000 11297
100.0 %   424.89156413207047 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56739), packet_size 700
966 676200 11040
96.6 %   490.0 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56740), packet_size 800
931 744800 13054
93.1 %   456.4424697410755 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56741), packet_size 900
983 884700 12458
98.3 %   568.1168726922459 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56742), packet_size 1000
961 961000 20034
96.1 %   383.7476290306479 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56743), packet_size 1100
927 1019700 23524
92.7 %   346.77775888454346 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56744), packet_size 1200
876 1051200 16513
87.6 %   509.27148307394174 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56745), packet_size 1300
883 1147900 19798
88.3 %   463.84483281139507 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56746), packet_size 1400
883 1236200 19543
88.3 %   506.0430844803766 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56747), packet_size 1500
986 1479000 31836
98.6 %   371.6547304937806 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56748), packet_size 1600
998 1596800 38297
99.8 %   333.56137556466564 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56749), packet_size 1700
939 1596300 35454
93.9 %   360.1963107124725 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56750), packet_size 1800
994 1789200 35618
99.4 %   401.86422595316975 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56751), packet_size 1900
999 1898100 35385
99.9 %   429.1309877066554 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56752), packet_size 2000
977 1954000 34803
97.7 %   449.15668189523893 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56753), packet_size 3000
961 2883000 58102
96.1 %   396.957075487935 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56754), packet_size 4000
990 3960000 69556
99.0 %   455.46034849617575 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56755), packet_size 5000
982 4910000 80850
98.2 %   485.8379715522573 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56756), packet_size 6000
966 5796000 101974
96.6 %   454.7041402710495 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56757), packet_size 7000
991 6937000 102771
99.1 %   539.9966916737212 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56758), packet_size 8000
998 7984000 125969
99.8 %   507.0453841818225 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56759), packet_size 9000
1009 9081000 143361
100.9 %   506.74869734446605 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56760), packet_size 10000
998 9980000 149575
99.8 %   533.779040615076 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56761), packet_size 11000
999 10989000 167271
99.9 %   525.5662966084976 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56762), packet_size 12000
976 11712000 183157
97.6 %   511.56111969512494 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56763), packet_size 13000
991 12883000 192345
99.1 %   535.8288492032545 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56764), packet_size 14000
969 13566000 208836
96.9 %   519.6805148537609 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56765), packet_size 15000
925 13875000 224239
92.5 %   495.0075588992102 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56766), packet_size 16000
990 15840000 232307
99.0 %   545.4850693263655 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56767), packet_size 17000
959 16303000 249645
95.9 %   522.4378617637044 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56768), packet_size 18000
935 16830000 272024
93.5 %   494.956327382878 MBit/s
waiting... Local IP address: 192.168.4.177
connect ('192.168.4.199', 56769), packet_size 19000
989 18791000 279460
98.9 %   537.9231374794246 MBit/s
waiting... Local IP address: 192.168.4.177
```

Note that this is via Ethernet from the camera to the PC via WiFi.

kwagyeman commented:
With the 100Mb PoE Shield:

PC: `iperf3.py -s`

PyBoard `client('192.168.4.177', udp=False, reverse=False)`

```
 0.00-10.00  sec  29.8 MBytes  25.0 Mbits/sec  sender
 --- SOME ISSUE CAUSES THE TEST NOT TO END ---
```

Not sure why, but, the above mode test never exits with ipref3.py.

PyBoard: `client('192.168.4.177', udp=True, reverse=False)`

```
  0.00-60.00  sec  75.0 MBytes  10.5 Mbits/sec  53957  sender
  0.00-60.08  sec  74.9 MBytes  10.5 Mbits/sec  53902  receiver
```

PyBoard: `client('192.168.4.177', udp=False, reverse=True)`

```
  0.00-60.00  sec   160 MBytes  22.4 Mbits/sec  sender
  0.00-60.05  sec   160 MBytes  22.3 Mbits/sec  receiver
```

PyBoard `client('192.168.4.177', udp=True, reverse=True)`

PC errors out with something about not being able to bind sockets.

kwagyeman commented:
So, things appear to work with both the 100Mb and 1Gb PoE shields. 

Client or Server mode with UDP and Reverse True doesn't seem to work using either the PC or the Board.

And, there's some issue with the default cilent TCP mode not exiting when printing receiver stats. Not sure if this happens for you.

...

I see the activity LED blinking too!

---

@dpgeorge - If the above issues are just with the ipref3 python script then please finish the PR and merge.

> I would be very interested to know how you managed 800Mbit/sec. Can you really achieve this at the other side of the link, receiving UDP packets at that rate? How long was the test run that measured 800Mbit/sec? It's possible to fill the DMA very quickly (greater than 1Gbit/sec), but it can take longer for it to drain.

This was what ST's ipref test reported. Note that there was a 0% packet reception rate on the PC. It was just the number the board published.

kwagyeman commented:
Yes, but the higher-end test already confirms the max performance. ipref testing was just to ensure all systems work. :)

dpgeorge commented:
> If the above issues are just with the ipref3 python script then please finish the PR and merge.

That's for the thorough testing.  Using the WLAN on the OpenMV N6 I could reproduce the issue when running `ipref3.py -c -R -u` on the PC, so that's most likely related to the `iperf3.py` script itself.  But the other modes work with WLAN so I'll see if there's something wrong there with the LAN code.

kwagyeman commented:
Yes, it would be this mode:

PC: iperf3.py -s

PyBoard client('192.168.4.177', udp=False, reverse=False)

```
0.00-10.00  sec  30.1 MBytes  25.2 Mbits/sec  sender
 --- SOME ISSUE CAUSES THE TEST NOT TO END - HANG ON PRINTING RECEIVER STATS ---
```

andrewleech commented:
I've been trying to make a C extension to iperf3 to test speed directly at the lwip / C level to see if the python socket interface / VM is adding much overhead, but having tricky boot issues. 

I'm finding when adding a user C module to the build, if it has three attributes / functions exposed in its module global dict it's fine, but if I add a fourth the board will not boot (IACCVIOL faults), similarly if I try to add qwargs to one of the functions.

Have your seen things like this? Maybe related to the qstr table size or something?

kwagyeman commented:
@andrewleech - Odd. Not sure what's wrong. I ran the scripts with OpenMV's firmware and IDE.

dpgeorge committed:
stm32/eth: Move DMA TX/RX index variables to normal RAM.

RAM shared with DMA needs caching disabled, which is slower for the MCU to
access.  So move these index variables (which are not shared with DMA) to
normal RAM.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
@andrewleech to improve the `udp_perf.py` script's reliability, you can try adding a `time.sleep(0.2)` at the end of the `client()` function.  That allows time for "stray" UDP packets to be sent/received and they don't mess up the subsequent test run.
